dhall/src/Dhall/Binary.hs

```diff
@@ -281,12 +282,13 @@ instance ToTerm a => ToTerm (Expr s a) where
         l₁ = encode l₀
         r₁ = encode r₀
     encode (ListLit _T₀ xs₀)
-        | null xs₀  = TList [ TInt 4, _T₁ ]
+        | null xs₀  = TList [ TInt label, _T₁ ]
         | otherwise = TList ([ TInt 4, TNull ] ++ xs₁)
       where
-        _T₁ = case _T₀ of
-            Nothing -> TNull
-            Just t  -> encode t
+        (label, _T₁) = case fmap Dhall.Core.denote _T₀ of
```

Comment:
`denote` was required to get the parser tests to pass. I suspect the denoting should happen in the test setup though.

Usually we can assume that the expressions we encode here are already denoted, right?

Reply:
We can enforce that in the types by changing the first type parameter of `Expr` to `X`

Reply:
You mean changing the instance head to

```
instance ToTerm a => ToTerm (Expr X a)
```

?

I tried to do this and followed the type errors for a bit. Eventually there's some code in `Dhall.Import` that currently relies on an instance for `Expr Src X`. Maybe I'll revisit that idea later.

Reply:
I realize now that the `denot`ing here is only necessary because I stopped `shallowDenot`ing the type in `Dhall.Parser.Expression`. I think the parser change was a good idea as we now have a more fully annotated parse tree.

Is it ok if I leave this code like it is now?

Reply:
If we still want to change the `ToTerm` instance, we should do that separately, as that will be a good-sized change by itself.

Reply:
I think you should either (A) `denote` the entire syntax tree once before decoding, rather than doing a `denote` here or (B) do a `shallowDenote` here.  The main thing that I want to prevent is the inefficiency of multiple recursive `denote`s

Personally, I think (A) is simpler

Reply:
(A) would mean that an already denoted Expr would get denoted again when we encode it. I'd rather change the instance head from `Expr s a` to `Expr X a`.

But that might cause some churn elsewhere – so I'm not too keen on it.

I think the `denote` here is ok: `denote` is very lazy, so it will only denote the `List` and the `App`, and be done.

If you disagree,  I'd move `shallowDenote` into `Dhall.Core` and re-use it here.

Reply:
In that case I think we should go with `shallowDenote`

Even though `denote` is lazy, if you have nested lists they will build up (i.e. the Nth nested list would have to step through N separate `denote` functions for each node in the abstract syntax tree).  I know that this is unlikely to be an issue in practice, but it can't hurt to fix